### ISSUE TITLE
Single Blind Option for Princess

### ISSUE THREAD

**Issue by scJazz2** [MegaMek/megamek#203]
Reference... https://sourceforge.net/p/megamek/feature-requests/1354/

after chatting with Arlith, Akjosch, and others on Esper #battletech...

I would like to RFE a Single Blind option for Princess games. Princess does a /seeall for each Princess player? Human plays as Blind.

**Comment by pheonixstorm** [comment]
I would rather see princess fixed to allow for double blind myself. It makes more sense than to handicap the player and allow the bot to cheat.

**Comment by scJazz2** [comment]
low hanging fruit dude... low hanging. It is at least a start

**Comment by pheonixstorm** [comment]
Yeah, but it would still be a lot of work compared to just fixing princess which would probably just be harder but take less time. If you do single blind then later on fix princess you have to still remove the old code. But, that's just me looking at it from a programmers perspective rather than a players.

**Comment by BLOODWOLF333** [comment]
Single blind already kinda works.  /seeall 1 (if princess is #1 under /who) and that commands MM to give her full vision while you are still under the blind rules.  AJ had mentioned something in passing about thats how AI's in other games do it as well where the bot always knows where the player is the whole time IIRC.

**Comment by pheonixstorm** [comment]
Yeah, it is one of the AI cheats for almost ALL RTS/FPS games as a means to keep the AI coding down since AI has never been a high priority for games.

**Comment by dericpage** [comment]
Single blind would not only be easier to code but it also would not need to be intrusive.  What Princess can or cannot see is not housed within the Princess code at all.  That's handled by the game engine itself.  So, if Princess gets updated to "properly" handle double-blind play, there's no need to remove any of the single-blind code.  The option would simply need to be turned off.

**Comment by BLOODWOLF333** [comment]
Was thinking about this and wondered if to simulate double-blind with princess that at the beginning of the game princess would have the locations of all the players units after they all deploy.  Then unless she has LOS to the players units she would go off the assumption that all the units are still where the player deployed them(last known location) and would receive an update on all the player unit locations every six turns (1 min game time).  That way she would at least go in the right direction even if going after ghost targets until the one minute update occurs but, she wouldn't always know each player unit's exact location 100% of the time.

**Comment by pheonixstorm** [comment]
Interesting idea, worth looking into I think.

**Comment by scJazz2** [comment]
Hmmm,.. that does of course mean fiddling with Princess instead of the Visibility code. Still interesting idea.

**Comment by dericpage** [comment]
You could do an option that, once every X turns, turns on Single Blind for one turn and then turns it off again.

**Comment by SirMegaV** [comment]
How about just for round 1 so that Princess have the last known location
and move towards the right direction? But whatever the solution is, the
documentation/remarks must be added clearly so that in the future of when
the Princess is enchanted, this solution will also get removed.

**Comment by scJazz2** [comment]
How about we just leave everything well enough alone and just implement the original, simple, no fuss, request and then after that we look at making things more complicated?

**Comment by Akjosch** [comment]
+1 to simply setting bot players to see everything via an option. It's essentially a single flag set at game start, two to three lines of code, plus some for the option.

In addition, entity evaluation for Princess happens in `megamek.client.bot.princess.BasicPathRanker``, lines 496ff. Simply lowering the maximum damage done and taken by an enemy to half or one-third for technically "invisible" enemies when in double-blind play would mean Princess will tend to ignore those most of the time. Another few lines of code improving things.

**Comment by arlith** [comment]
I don't think the /seeAll option will work as a workaround.  This relies upon `Player.canSeeAll()`, which checks to see if a player is an observer *and* has the see_entire_board state set.  Not to say that this couldn't be modified for Princess, but just using `/seeAll <Princess-PID>` won't work as expected.

**Comment by BLOODWOLF333** [comment]
I tested it when connected as a computer player and it gives princess full vision but, never got around to testing it inside one instance of MM vs the bot.

**Comment by pheonixstorm** [comment]
I did and princess found me fairly quickly, though it could have been a fluke. Will test some more in the next battle I have. Along with more information (map, terrain, etc.)

**Comment by arlith** [comment]
It's possible I missed something, or maybe Princess counts as an observer.

**Comment by scJazz2** [comment]
Nothing more really needs to be tested I think. If Princess can't see you she just herds in place.

**Comment by arlith** [comment]
The `/seeAll` command works by setting the `Player.see_entire_board` state.  This is used in two methods in `Player`: `canSeeAll()` and `getSeeAll()`.  The latter just returns the value of `Player.see_entire_board`, and is only used for artillery attacks.  `Player.canSeeAll()` is used in the primary double-blind related methods (`Server.filterEntities`, `Server.whoCanSee`, and `Entity.hasSeenEntity`), however it returns `Player.observer && Player.see_entire_board`, which means that it only takes effect if the player is also an observer (to prevent cheating).

For the `/seeAll` command to have an effect, the bot player would need to  be an observer, and players controlled by bots are no different from any other player (bots are implemented as clients that play for a player).

This framework could be used to create a singleblind mode for Princess, but from my interpretation of the code, `/seeAll` won't do anything right now.  Of course, I didn't run any actual tests, just looked at the code...

**Comment by BLOODWOLF333** [comment]
Just tested this in just a single window of MM, no connecting as a computer player in another.  Single Blind already works.  1x1 map, deployed my mech behind hills where it was hidden.  /seeall 1 and princess does come and get me.  Had clicked through a few turns at the start but, as soon as i sent the command she found me.
[sb.sav.gz](https://github.com/MegaMek/megamek/files/296911/sb.sav.gz)
player name: BLOODWOLF

**Comment by arlith** [comment]
Interesting.  I'll have to see why it's working, because it's not clear from the code I saw that it should work.  Ideally, I think this should be an option in Princess Behavior, so that it's more visible.  That was what I was hoping to do  to close this ticket, but then when I looked at the code it appeared that /seeall shouldn't work.

**Comment by BLOODWOLF333** [comment]
Alright, Arlith is right.  I created a map to test this drawHex NPE that has mountains and buildings and every terrain type.  /seeall 1 to give princes full vision and single blind no longer works.  It never worked.  She knows to pursue sensor contacts which are the same as full detection to her.  So, we all were just under the illusion that single blind works because previously the map I had used it on were not very severe terrain-wise.  This current map, once I get ontop of the hill and can see her I can see that she hasnt moved her units at all because is receiving no sensor contacts from my mechs with them behind a mountain and buildings.
